flow/layers/image_filter_layer.cc

```diff
@@ -11,6 +11,33 @@ ImageFilterLayer::ImageFilterLayer(sk_sp<SkImageFilter> filter)
       transformed_filter_(nullptr),
       render_count_(1) {}
 
+void ImageFilterLayer::Diff(DiffContext* context, const Layer* old_layer) {
+  auto subtree = context->BeginSubtree();
+  auto* prev = reinterpret_cast<const ImageFilterLayer*>(old_layer);
+  if (!context->IsSubtreeDirty()) {
+    assert(prev);
+    if (filter_ != prev->filter_) {
+      context->MarkSubtreeDirty(old_layer->paint_region());
+    }
+  }
+
+  DiffChildren(context, prev);
+
+  SkMatrix inverse;
+  if (context->GetTransform().invert(&inverse)) {
+    auto paint_bounds = context->CurrentSubtreeRegion().GetBounds();
+    inverse.mapRect(&paint_bounds);
+    paint_bounds = filter_->computeFastBounds(paint_bounds);
+
+    if (context->IsSubtreeDirty()) {
+      context->AddPaintRegion(paint_bounds);
```

Comment:
Don't we need to `context->AddPaintRegion(paint_bounds);` no matter whether the subtree is dirty or not?

Reply:
No, that would means image filter gets repainted on every frame whether it needs to or not. The line next to it:
```cpp
context->AddReadbackRegion(paint_bounds);
```
basically means that if anything within the paint bounds of image filter needs to be repainted, the whole image filter will be repainted. (we can't partially repaint image filter because the filter can potentially sample out of clip).